### PR TITLE
[ci] enable ruff-format on some files, add pre-commit config

### PR DESCRIPTION
**Pull request by jameslamb**
Contributes to #6304.

* adds auto-formatting of Python code with `ruff format`
* moves enforcement of that formatting, `ruff` linter checks, and `isort` into a `pre-commit` config
* updates `CONTRIBUTING.md` with information on how to do that

### Notes for Reviewers

The versions of tools I added in `.pre-commit-config.yaml` are the latest released versions available.

As discussed over in #6304, in this PR I've omitted the directories with a lot of Python code (`examples/`, `python-package/`, and `tests/`). So we can focus on the types of changes that'll be made and just see a few small examples in other code.

Also as described in #6304, I'll add the commit from this PR (and the others that reformat the code) in a `.git-blame-ignore-revs` file once we're done with this work.

### CODE REVIEW

File: .ci/test.sh [hunk 1]
```diff
@@ -74,10 +74,9 @@ if [[ $TASK == "lint" ]]; then
         ${CONDA_PYTHON_REQUIREMENT} \
         cmakelint \
         cpplint \
-        isort \
```
Comment (jameslamb) [inline .ci/test.sh:77]: `pre-commit` installs packages into its own virtualenv, so it's not necessary to pre-install them any more

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,87 @@
+[tool.isort]
+known_first_party = [
+    "lightgbm"
+]
```
Comment (jameslamb) [inline pyproject.toml:4]: This one was interesting! Once this configuration was moved out of the `pyproject.toml` with a `[project]` block, `isort` treated `lightgbm` as just another third-party import and moved around all the imports like `import lightgbm as lgb`.

See https://pycqa.github.io/isort/docs/configuration/options.html#known-first-party.

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    "examples/*.py",
+    "external_libs/*.py",
+    "lightgbm-python/*.py",
+    "python-package/*.py",
+    "tests/*.py"
+]
+indent-style = "space"
+quote-style = "double"
+
+[tool.ruff.lint]
```
Comment (jameslamb) [inline pyproject.toml:37]: The introduction of this table (and some other renamings) are in response to this warning from `ruff`:

```text
warning: The top-level linter settings are deprecated in favour of their counterparts in the `lint` section. Please update the following options in `pyproject.toml`:
  - 'ignore' -> 'lint.ignore'
  - 'select' -> 'lint.select'
  - 'pydocstyle' -> 'lint.pydocstyle'
  - 'per-file-ignores' -> 'lint.per-file-ignores'
```

Which have been showing up in the `lint` CI job, e.g. https://github.com/microsoft/LightGBM/actions/runs/7852675526/job/21431398282#step:3:405

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  )$
+
+repos:
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    # Ruff version.
+    rev: v0.2.1
+    hooks:
+      # Run the linter.
+      - id: ruff
+        args: ["--config", "pyproject.toml"]
```
Comment (borchero) [inline .pre-commit-config.yaml:16]: I think this is optional. Instead, I would maybe add configuration to autofix auto-fixable issues:
```suggestion
        args: ["--fix", "--exit-non-zero-on-fix"]
```

File: pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    # (flake8-bugbear) Found useless expression
+    "B018",
+    # pydocstyle
+    "D",
+    # flake8-print
+    "T"
+]
+
+[tool.ruff.lint.pydocstyle]
+
```
Comment (borchero) [inline pyproject.toml:86]: [nit]

```suggestion
```

**Review by borchero — APPROVED** [review]
Looks reasonable to me! Thanks for setting this up :)

File: pyproject.toml (hunk 5 above)
Comment (jameslamb) [reply pyproject.toml:86]: ah you're right, I was inconsistent here. Thanks!

File: .pre-commit-config.yaml (hunk 4 above)
Comment (jameslamb) [reply .pre-commit-config.yaml:16]: I agree that it's optional, and `pyproject.toml` is picked up automatically. Leaving this in was a mistake. I think I'd had that at some point while trying to figure out how to have different `exclude` patterns for `ruff check` vs. `ruff format`.

I'll remove this.

However... I don't support adding `--fix` for `ruff check`. I'd rather have the opportunity to decide for myself whether to change something or to ignore it like this:

https://github.com/microsoft/LightGBM/blob/cc733f8595267f313886f92ed5d1285010ba8f3f/python-package/pyproject.toml#L131-L145

I've worked in other projects with `--fix` and found it annoying when `pre-commit` undoes my not-yet-committed work in significant ways (like removing variables that appear "unused" but which I just haven't yet finished the implementation for).

I think we should omit `--fix`.

Comment (jameslamb) [reply .pre-commit-config.yaml]: I just pushed https://github.com/microsoft/LightGBM/pull/6308/commits/4f1a3d3f154820f23f93a52e0d8503baecefc138 removing this unnecessary `--config`, thanks for catching that!

Comment (borchero) [reply .pre-commit-config.yaml]: > I think we should omit `--fix`.

Fine for me ;)

**Comment by jmoralez** [comment]
I think having two `pyproject.toml` files is a bit confusing. If we're going to replace isort with ruff in the future it may be worth to define a `ruff.toml` and a temporal `.isort.cfg` instead. WDYT?

**Comment by jameslamb** [comment]
> I think having two `pyproject.toml` files is a bit confusing. If we're going to replace isort with ruff in the future it may be worth to define a `ruff.toml` and a temporal `.isort.cfg` instead. WDYT?

I don't agree that it's confusing.

* project-wide configuration at the top-level `pyproject.toml`
* details for building the `lightgbm` Python package in `python-package/pyproject.toml`

I'd prefer to not have a proliferation of tool-specific config files at the root of the repo. And I like having it at the root of the repo so that if you just call `ruff check` or `isort` or `mypy` or whatever directly from the root of the repo (not using `pre-commit`), you'll still automatically pick up the configuration used in CI.

However, I feel most strongly about not having the tool-specific config files. So if you don't like having multiple `pyproject.toml`, would you support adding stuff like this in `.pre-commit-config.yaml`?

```yaml
 args: ["--config", "python-package/pyproject.toml"]
```

And moving the configuration back into there?

**Comment by jmoralez** [comment]
> would you support adding stuff like this in .pre-commit-config.yaml?
`args: ["--config", "python-package/pyproject.toml"]`
And moving the configuration back into there?

Yes, I think that's clearer

**Comment by jameslamb** [comment]
Alright no problem, I'm good with that! Just pushed a commit moving the settings back into `python-package/pyproject.toml` and telling `pre-commit` to look there for configuration.

**Review by jmoralez — APPROVED** [review]
Thanks!

**Pull request merged**